### PR TITLE
Mitigate node out of disk status oscillation.

### PR DESCRIPTION
**Pull request by madhusudancs**
Fix for issue #12041.

### CODE REVIEW

**Comment by madhusudancs** [comment]
Did not know Github allows only one assignee per PR.

@davidopp @lavalamp PTAL.

**Comment by davidopp** [comment]
IIUC this PR will delay reporting
nodeOODCondition.Status = api.ConditionTrue
if you transitioned to reporting
nodeOODCondition.Status = api.ConditionFalse
less than 5 minutes ago. (And also the opposite transition.)

But I think this is not good -- I agree we want to delay reporting
nodeOODCondition.Status = api.ConditionFalse
if we transitioned to reporting
nodeOODCondition.Status = api.ConditionTrue
less than 5 minutes ago, but I do not think we ever want to 
delay reporting
nodeOODCondition.Status = api.ConditionTrue

In other words, I think we want to report "out of disk" as soon as it happens, so scheduler can immediately start avoiding that node, and only delay reporting "not out of disk." Does this make sense?

I think this is a simple change, just move the 
`if kl.clock.Since(nodeOODCondition.LastTransitionTime.Time) >= kl.outOfDiskTransitionFrequency {`
test into the branch that sets nodeOODCondition.Status = api.ConditionFalse
(and also update any relevant comments and flag description).

File: docs/admin/kubelet.md [hunk 1]
```diff
@@ -113,6 +113,7 @@ kubelet
       --network-plugin-dir="/usr/libexec/kubernetes/kubelet-plugins/net/exec/": <Warning: Alpha feature> The full path of the directory in which to search for network plugins
       --node-status-update-frequency=10s: Specifies how often kubelet posts node status to master. Note: be cautious when changing the constant, it must work with nodeMonitorGracePeriod in nodecontroller. Default: 10s
       --oom-score-adj=-999: The oom-score-adj value for kubelet process. Values must be within the range [-1000, 1000]
+      --outofdisk-transition-frequency=5m0s: Duration for which the kubelet has to wait before transitioning from one out of disk node condition status to another. Default: 300s
```
Comment (davidopp) [inline docs/admin/kubelet.md]: elsewhere you wrote 300s

Comment (madhusudancs) [inline docs/admin/kubelet.md]: Made it 300s.

**Comment by madhusudancs** [comment]
@davidopp Addressed the comment. PTAL.

**Comment by madhusudancs** [comment]
@k8s-bot test this please.

File: docs/admin/kubelet.md [hunk 2]
```diff
@@ -113,6 +113,7 @@ kubelet
       --network-plugin-dir="/usr/libexec/kubernetes/kubelet-plugins/net/exec/": <Warning: Alpha feature> The full path of the directory in which to search for network plugins
       --node-status-update-frequency=10s: Specifies how often kubelet posts node status to master. Note: be cautious when changing the constant, it must work with nodeMonitorGracePeriod in nodecontroller. Default: 10s
       --oom-score-adj=-999: The oom-score-adj value for kubelet process. Values must be within the range [-1000, 1000]
+      --outofdisk-transition-frequency=5m0s: Duration for which the kubelet has to wait before transitioning out of out-of-disk node condition status. Default: 300s
```
Comment (davidopp) [inline docs/admin/kubelet.md]: I would suggest doing 300s instead of 5m0s just for consistency.

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -320,7 +320,8 @@ func NewMainKubelet(
 		resolverConfig:                 resolverConfig,
 		cpuCFSQuota:                    cpuCFSQuota,
 		daemonEndpoints:                daemonEndpoints,
-		containerManager:               containerManager,
+		clock:                          util.RealClock{},
+		outOfDiskTransitionFrequency: outOfDiskTransitionFrequency,
```
Comment (davidopp) [inline pkg/kubelet/kubelet.go]: this appears to be unaligned; you probably need to run gofmt

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -637,6 +638,17 @@ type Kubelet struct {
 
 	// oneTimeInitializer is used to initialize modules that are dependent on the runtime to be up.
 	oneTimeInitializer sync.Once
+
+	// clock is an interface that provides time related functionality in a way that makes it
+	// easy to test the code.
+	clock util.Clock
+
+	// outOfDiskTransitionFrequency specifies the duration for which the kubelet has to wait
```
Comment (davidopp) [inline pkg/kubelet/kubelet.go]: This comment seems a bit overly verbose. I would say something like "outOfDiskTransitionFrequency specifies the amount of time the kubelet has to be actually not out of disk before it can transition the node condition status from out-of-disk to not-out-of-disk. This prevents a pod that causes out-of-disk condition from repeatedly getting rescheduled onto the node."

File: pkg/kubelet/kubelet.go [hunk 5]
```diff
@@ -2731,7 +2741,7 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
 			kl.recordNodeStatusEvent("NodeOutOfDisk")
 		}
 	} else {
-		if nodeOODCondition.Status != api.ConditionFalse {
+		if nodeOODCondition.Status != api.ConditionFalse && kl.clock.Since(nodeOODCondition.LastTransitionTime.Time) >= kl.outOfDiskTransitionFrequency {
```
Comment (davidopp) [inline pkg/kubelet/kubelet.go]: maybe log a message if
nodeOODCondition.Status != api.ConditionFalse
but the clock condition is not true yet?

Comment (davidopp) [inline pkg/kubelet/kubelet.go]: I think we should skip the clock check if nodeOODCondition.Status == api.ConditionUnknown. If we don't do that, I think on startup we will not start reporting nodeOODCondition.Status == api.ConditionFalse for five minutes, since the field is initialized to api.ConditionUnknown at startup. Same problem if the node is temporarily unreachable and NodeController marks it as nodeOODCondition.Status == api.ConditionUnknown. Do you agree?

File: pkg/kubelet/kubelet_test.go [hunk 6]
```diff
@@ -2926,6 +2993,166 @@ func TestUpdateExistingNodeStatus(t *testing.T) {
...
+		}
+		updatedNode, ok := updateAction.GetObject().(*api.Node)
+		if !ok {
+			t.Errorf("%d. unexpected object type", tcIdx)
+		}
+		kubeClient.ClearActions()
+
+		var oodCondition api.NodeCondition
+		for i, cond := range updatedNode.Status.Conditions {
+			// Expect LastProbeTime to be updated to Now, while LastTransitionTime to be the same.
```
Comment (davidopp) [inline pkg/kubelet/kubelet_test.go]: This comment should be "Expect LastHeartbeatTime to be updated to Now"

File: pkg/kubelet/kubelet_test.go [hunk 7]
```diff
@@ -2926,6 +2993,166 @@ func TestUpdateExistingNodeStatus(t *testing.T) {
...
+			if tcIdx == len(testCases)-1 && cond.Type == api.NodeOutOfDisk {
+				if clock.Since(cond.LastTransitionTime.Time) > 0 {
+					t.Errorf("%d. Condition %v LastTransitionTime:\nexpected \n%#v\n, got \n%#v", tcIdx, cond.Type, clock.Now(), cond.LastTransitionTime.Time)
+				}
+			} else {
+				if clock.Since(cond.LastTransitionTime.Time) == 0 {
+					t.Errorf("%d. Condition %v LastTransitionTime:\nexpected \n%#v\n, got \n%#v", tcIdx, cond.Type, clock.Now(), cond.LastTransitionTime.Time)
+				}
+			}
+			updatedNode.Status.Conditions[i].LastHeartbeatTime = unversioned.Time{}
```
Comment (davidopp) [inline pkg/kubelet/kubelet_test.go]: I'm sure I'm missing someting, but I didn't understand the purpose of this line and the next line. Don't we want LastHeartbeatTime and LastTransitionTime to be set by the Kubelet, not by the test itself?

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (madhusudancs) [inline pkg/kubelet/kubelet.go]: I think this is a gofmt bug. Since gofmt is run every time I save this file in the editor, this strange alignment keeps coming back. I even tried opening this file in an editor without gofmt integration, manually adjusting the alignment and then running gofmt on the command line again. This unaligned formatting just keeps happening.

File: pkg/kubelet/kubelet.go (hunk 4 above)
Comment (madhusudancs) [inline pkg/kubelet/kubelet.go]: Done.

File: pkg/kubelet/kubelet.go (hunk 5 above)
Comment (madhusudancs) [inline pkg/kubelet/kubelet.go]: > maybe log a message if nodeOODCondition.Status != api.ConditionFalse but the clock condition is not true yet?

Done.

> I think we should skip the clock check if nodeOODCondition.Status == api.ConditionUnknown. If we don't do that, I think on startup we will not start reporting nodeOODCondition.Status == api.ConditionFalse for five minutes, since the field is initialized to api.ConditionUnknown at startup. Same problem if the node is temporarily unreachable and NodeController marks it as nodeOODCondition.Status == api.ConditionUnknown. Do you agree?

Agreed. And fixed.

File: pkg/kubelet/kubelet_test.go (hunk 6 above)
Comment (madhusudancs) [inline pkg/kubelet/kubelet_test.go]: Done.

File: pkg/kubelet/kubelet_test.go (hunk 7 above)
Comment (madhusudancs) [inline pkg/kubelet/kubelet_test.go]: A few lines below, we do a deep equals between the objects. It is hard to compare time values.  So I followed this pattern that other tests use in this file.

Now that we can precisely control the time set by the tests, we can switch to deep equals based comparisons even for time values.

File: docs/admin/kubelet.md (hunk 2 above)
Comment (madhusudancs) [inline docs/admin/kubelet.md]: I tried changing this before, but then I realized that this is an auto-generated file. It seems to parse 300s and automatically write it out as 5m0s. I have changed all the relevant occurrences to 5m0s for consistency.

**Comment by madhusudancs** [comment]
@davidopp PTAL.

**Comment by vishh** [comment]
Given the fact that kubernetes already has low and high thresholds for disk reclamation, why do we think that kubelet will oscillate? 
The current defaults as percentage of disk capacity are 90% for high threshold and 60% for low threshold.
Kubelet attempts to reclaim disk space until it hits the low watermark.

**Comment by davidopp** [comment]
Needs rebase.

@vishh The scenario we're concerned about is a small machine where we try to schedule a container with a large image that causes the machine to immediately go OOD; so the pod fails, and the machine immediately goes back to not-OOD once the image is GCd, so the scheduler re-tries that pod on that same machine again, etc. in an infinite loop. #12041 has more details.

**Comment by davidopp** [comment]
LGTM

**Comment by davidopp** [comment]
(but needs rebase)

**Comment by madhusudancs** [comment]
Rebased. PTAL.

**Comment by davidopp** [comment]
The changes to the files in docs/api-reference look suspicious. Were those intentional?

**Comment by davidopp** [comment]
Also, please squash the commits.

**Comment by madhusudancs** [comment]
> The changes to the files in docs/api-reference look suspicious. Were those intentional?

It wasn't intentional, it was a result of a rebase. Regenerated those files ago.

Squashed the commits. PTAL.

**Comment by davidopp** [comment]
Wow, that's annoying that the auto-generators produce so much extra change to a PR.

Anyway, I've marked LGTM. Thanks!

File: pkg/kubectl/testing/types.generated.go [hunk 8]
```diff
@@ -158,7 +158,13 @@ func (x *TestStruct) CodecEncodeSelf(e *codec1978.Encoder) {
 				z.EncSendContainerState(codecSelfer_containerArrayElem1234)
 				if yyq2[2] {
 					yy10 := &x.ObjectMeta
-					yy10.CodecEncodeSelf(e)
+					yym11 := z.EncBinary()
```
Comment (wojtek-t) [inline pkg/kubectl/testing/types.generated.go]: Why do you change those auto-generated types? Your changes seem incorrect and are reason for test failures...

Comment (madhusudancs) [inline pkg/kubectl/testing/types.generated.go]: I did not change these. This is what hack/update-all.sh generated. And the pre-commits were passing.

Comment (madhusudancs) [inline pkg/kubectl/testing/types.generated.go]: So I ran hack/update-all.sh again, which runs hack/update-codecgen and I don't get any more diffs. I am not entirely sure what is going on.

**Comment by madhusudancs** [comment]
@davidopp For an unrelated experiment, I had upgraded Go to 1.5.2 and that likely caused those go2idl testdata types to be generated. A similar behavior has also been reported in #18907. I downgraded Go to 1.4.3 and the generated code looks Ok now. I have also rebased the PR again. PTAL.

**Comment by davidopp** [comment]
LGTM

I'm guessing the test failures are related to the problems we're having with the tests lately and not with this PR, but I didn't check.

**Comment by madhusudancs** [comment]
@k8s-bot test this please.

**Comment by madhusudancs** [comment]
@k8s-bot test this please.

**Comment by madhusudancs** [comment]
@k8s-bot test this please.
